In [1]:
# GPUのアサイン確認
import tensorflow as tf
tf.test.gpu_device_name()

ModuleNotFoundError: No module named 'tensorflow'

In [2]:
# プログレスバーとMeCabのインストール
# MeCab 参考サイト：https://news.mynavi.jp/article/zeropython-26/
!pip install tqdm
!pip install ipywidgets # プログレスバーの表示で必要
!apt-get install mecab libmecab-dev mecab-ipadic-utf8
!pip install mecab-python3

Invalid requirement: '#'
Traceback (most recent call last):
  File "c:\users\admin\anaconda3\lib\site-packages\pip\_vendor\packaging\requirements.py", line 93, in __init__
    req = REQUIREMENT.parseString(requirement_string)
  File "c:\users\admin\anaconda3\lib\site-packages\pip\_vendor\pyparsing.py", line 1632, in parseString
    raise exc
  File "c:\users\admin\anaconda3\lib\site-packages\pip\_vendor\pyparsing.py", line 1622, in parseString
    loc, tokens = self._parse( instring, 0 )
  File "c:\users\admin\anaconda3\lib\site-packages\pip\_vendor\pyparsing.py", line 1379, in _parseNoCache
    loc,tokens = self.parseImpl( instring, preloc, doActions )
  File "c:\users\admin\anaconda3\lib\site-packages\pip\_vendor\pyparsing.py", line 3395, in parseImpl
    loc, exprtokens = e._parse( instring, loc, doActions )
  File "c:\users\admin\anaconda3\lib\site-packages\pip\_vendor\pyparsing.py", line 1379, in _parseNoCache
    loc,tokens = self.parseImpl( instring, preloc, doActions )
  File "

  Using cached https://files.pythonhosted.org/packages/25/e9/bbf5fc790a2bedd96fbaf47a84afa060bfb0b3e0217e5f64b32bd4bbad69/mecab-python3-0.7.tar.gz
    Complete output from command python setup.py egg_info:
    'mecab-config' は、内部コマンドまたは外部コマンド、
    操作可能なプログラムまたはバッチ ファイルとして認識されていません。
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "C:\Users\admin\AppData\Local\Temp\pip-install-kgjr35vl\mecab-python3\setup.py", line 41, in <module>
        include_dirs=cmd2("mecab-config --inc-dir"),
      File "C:\Users\admin\AppData\Local\Temp\pip-install-kgjr35vl\mecab-python3\setup.py", line 21, in cmd2
        return cmd1(strings).split()
      File "C:\Users\admin\AppData\Local\Temp\pip-install-kgjr35vl\mecab-python3\setup.py", line 18, in cmd1
        return os.popen(strings).readlines()[0][:-1]
    IndexError: list index out of range
    
    ----------------------------------------


Command "python setup.py egg_info" failed with error code 1 in C:\Users\admin\AppData\Local\Temp\pip-install-kgjr35vl\mecab-python3\


In [3]:
# 参考サイト：https://qiita.com/uni-3/items/201aaa2708260cc790b8

# drive内のディレクトリをマウントする
from google.colab import drive
drive.mount("/content/gdrive")

# 認証部分

# Install the PyDrive wrapper & import libraries.
# This only needs to be done once per notebook.
!pip install -U -q PyDrive
from pydrive.auth import GoogleAuth
from pydrive.drive import GoogleDrive
from google.colab import auth
from oauth2client.client import GoogleCredentials

# check auth
auth.authenticate_user()
gauth = GoogleAuth()
gauth.credentials = GoogleCredentials.get_application_default()
drive = GoogleDrive(gauth)

# 読み込み部分
downloaded = drive.CreateFile({'id':'1MDFE9qXab-tIs0dY1YexOMyKb8TP7r_C'})

# Download the file to a local disk as 'my_svr_training_kit.py'.
downloaded.GetContentFile('my_svr_training_kit.py')

Mounted at /content/gdrive


In [4]:
# drive内のディレクトリをマウントした後のディレクトリ内確認
!ls "/content/gdrive"

'My Drive'


In [5]:
# my_svr_training_kit.pyがある場所にチェンジディレクトリ
%cd "/content/gdrive/My Drive/sp_lecture_paper/src"

/content/gdrive/My Drive/sp_lecture_paper/src


In [1]:
## Google Colaboratoryで実行しない場合、このセルから実行すればOK
import my_svr_training_kit as my_stkit
from tqdm import tqdm_notebook as tqdm

In [2]:
# 学習データ:テストデータ = 8:2に分割したデータを取得する
X_train_norm, X_test_norm, y_train, y_test = my_stkit.prepare_traindata()

Finished loading input vector, bag of words: about 0.034023284912109375 seconds.
shape of bag of words vector: (9754, 6548)
Finished getting all user talks from dialogue_corpus.db: about 0.21415257453918457 seconds.
Finished making output vector, user acceptance labels: about 0.22716164588928223 seconds.
len(y) = 9754
訓練データ、交差検証データ、テストデータの数 = 5852 1951 1951
Finished splitting data: about 1.2739050388336182 seconds.

Finished standardization train_data and test_data: about 2.7639875411987305 seconds.
Finished preparing train_data of SVR.



In [2]:
# ホールドアウト法に基づき、学習とテストを行う
my_stkit.train_svr(X_train_norm, X_test_norm, y_train, y_test)

Finished loading input vector, bag of words: about 0.025017976760864258 seconds.
shape of bag of words vector: (9754, 6548)
Finished getting all user talks from dialogue_corpus.db: about 0.1961357593536377 seconds.
Finished making output vector, user acceptance labels: about 0.21215105056762695 seconds.
len(y) = 9754
訓練データ、交差検証データ、テストデータの数 = 5852 1951 1951
Finished splitting data: about 0.9317975044250488 seconds.

Finished standardization train_data and test_data: about 3.610603094100952 seconds.
Finished preparing train_data of SVR.

リッジ回帰
Finished fitting Ridge model: about 14.103132009506226 seconds.
accuracy of Ridge model: 0.051105226500055556, about 0.04303288459777832 seconds.
Finished saving Ridge model as trained_ridge_model.pkl: about 0.006005287170410156 seconds.

カーネル無し(線形カーネル)
Finished fitting Linear SVR model: about 8488.83792591095 seconds.
accuracy of Linear SVR model: -0.02302247791237666, about 144.44598054885864 seconds.
Finished saving Linear SVR model as trained_l

# ホールドアウト法による学習結果

## 各処理の所要時間や精度
※全体の所要時間：3時間程度(8時間放置してたら終わってた)

![SVRの学習結果](result_svr_training.jpg "学習結果")

### リッジ回帰

* モデルのフィッティング：14.103132009506226秒

* テストデータに対する精度：0.051105226500055556、計算にかかった時間：0.04303288459777832秒

* モデルの保存：0.006005287170410156秒

### カーネル無し(線形カーネル)

* モデルのフィッティング：8488.83792591095秒 = 2.35777777時間

* テストデータに対する精度：-0.02302247791237666(マイナス！？)、計算にかかった時間：144.44598054885864秒 = 2.407433009147643分

* モデルの保存：0.5323805809020996秒

### ガウシアンカーネル

* モデルのフィッティング：590.2907876968384秒 = 9.838179794947307分

* テストデータに対する精度：0.43980592965629134、計算にかかった時間：145.392639875412秒 = 2.4232106645902分

* モデルの保存：0.6056420803070068秒

## 追記：パラメータのチューニング

### リッジ回帰
分割数24で6分程度、分割数を8に減らしたら2分少々でパラメータのフィッティングが終わった。

### カーネル無し(線形カーネル)
分割数を8にして9時間放置しても終わらなかった。  
scikit-learnはGPUに対応していないからか、colaboratoryで2時間ほど放置しても終わらなかった。

### ガウシアンカーネル
線形カーネルがいつまで経っても終わらなかったため、所要時間を把握できていない。  
ただ、モデルの学習時間を見た感じ、線形カーネルよりはずっと短そう。

## 結論
ガウシアンカーネルのパラメータをチューニングさせるのが一番良さそう。

In [3]:
# ガウシアンカーネルのパラメータをチューニング
# -> チューニングしたパラメータでフィッティング
# -> テストデータに対する精度を計算
# -> チューニング済みのモデルを保存
my_stkit.tuning_rbf_kernel_svr(X_train_norm, X_test_norm, y_train, y_test)

ガウシアンカーネル
Finished tuning parameters: about 137291.3233382702 seconds.
C, εのチューニング
最適なパラメーター = {'C': 2.0691380811147897, 'epsilon': 0.42813323987193935}
精度 = 0.437369674666

Finished fitting with tuned parameters: about 258.87052607536316 seconds.

テストデータにフィット
テストデータの精度 = 0.443303032001

※参考
訓練データの精度 = 0.681647789927
交差検証データの精度 = 0.437369674666
Finished calculating accuracies: about 414.0194990634918 seconds.

Finished saving RBF SVR model as tuned_rbf_svr_model.pkl: about 0.33723950386047363 seconds.



# ガウシアンカーネル パラメータチューニング 結果

## パラメータの候補値 8個
※全体の所要時間：1時間程度

![ガウシアンカーネルのパラメータチューニングの結果(パラメータの候補値8個)](result_rbf_svr_tuning_20params.jpg "チューニング結果(パラメータの候補値8個)")

### パラメータのチューニング

* 所要時間：2151.963869333267秒 = 35.86606448888778分

* 最適なパラメーター = {'C': 1.0, 'epsilon': 0.10000000000000001}

* 精度 = 0.41851268794

### チューニングしたパラメータでのフィッティング

* 所要時間：355.791232585907秒 = 5.929853876431783分

* テストデータにフィット  
テストデータの精度 = 0.418710205889

* ※参考  
訓練データの精度 = 0.623500332303  
交差検証データの精度 = 0.41851268794  
精度の計算時間：551.9204204082489秒 = 9.198673673470815分

## パラメータの候補値 20個
※全体の所要時間：36時間程度(10月5日12:57～10月7日1:49)

![ガウシアンカーネルのパラメータチューニングの結果(パラメータの候補値20個)](result_rbf_svr_tuning_8params.jpg "チューニング結果(パラメータの候補値20個)")

### パラメータのチューニング

* 所要時間：137291.3233382702秒 = 38.13647870507506時間 = 1.589019946044794日

* 最適なパラメーター = {'C': 2.0691380811147897, 'epsilon': 0.42813323987193935}

* 精度 = 0.437369674666

### チューニングしたパラメータでのフィッティング

* 所要時間：258.87052607536316秒 = 4.314508767922718分

* テストデータにフィット  
テストデータの精度 = 0.443303032001

* ※参考  
訓練データの精度 = 0.681647789927  
交差検証データの精度 = 0.437369674666  
精度の計算時間：414.0194990634918秒 = 6.900324984391535分  

↓  
精度向上：パラメータ候補値を8個にして、二段グリッドサーチするしか無い・・・？

In [1]:
# 学習させたモデルを平均二乗誤差により評価

import my_svr_training_kit as my_stkit
from sklearn.externals import joblib
from sklearn.metrics import mean_squared_error
import numpy as np

# 学習データ:テストデータ = 8:2に分割したデータを取得する
X_train_norm, X_test_norm, y_train, y_test = my_stkit.prepare_traindata()

loaded_rbf_svr = joblib.load("tuned_rbf_svr_model.pkl")

pred_acceptance = loaded_rbf_svr.predict(X_test_norm)
true_acceptance = y_test
mse = mean_squared_error(true_acceptance, pred_acceptance)
rmse = np.sqrt(mse)

print("平均二乗誤差MSE = {}".format(mse))
print("平均二乗誤差RMSE = {}".format(rmse))
print()
print(pred_acceptance)
print()
print(true_acceptance)

Finished loading input vector, bag of words: about 0.02902054786682129 seconds.
shape of bag of words vector: (9754, 6548)
Finished getting all user talks from dialogue_corpus.db: about 0.19512057304382324 seconds.
Finished making output vector, user acceptance labels: about 0.2031261920928955 seconds.
len(y) = 9754
訓練データ、交差検証データ、テストデータの数 = 5852 1951 1951
Finished splitting data: about 1.3109374046325684 seconds.

Finished standardization train_data and test_data: about 3.226296901702881 seconds.
Finished preparing train_data of SVR.

平均二乗誤差MSE = 0.9462596970409166
平均二乗誤差RMSE = 0.9727588072286555

[ 1.42183811  2.9730976   2.52080883 ...,  2.83311699  1.92258212
  2.72433184]

[1 1 4 ..., 3 1 1]
